### ISSUE TITLE
Write Test for State Compiler when Package is Missing

### ISSUE THREAD

**Issue by rallytime** [saltstack/salt#27999]
In 2015.5.3 we had a regression in the state compiler, which took a while to realize how wide spread it was, where a state run that required an external dependency that hadn't been installed yet stacktraced instead of reporting the missing dependency. We need some tests written in our test suite to catch this so that this doesn't happen again.

Some of the issues that I found that exhibited this same stack trace and behavior are:
#26115
#26280
#26454
#26493
#26510
#26882
#27016
#26703
#27321
#27619

FYI @jfindlay, @jtand, and @ssgward.

**Comment by jfindlay** [comment]
Thanks for tracking these down @rallytime.  I saw some of them and thought there might be a pattern, but never investigated it enough to find the potential common connection.  https://github.com/saltstack/salt/issues/27016#issuecomment-148571302 may suggest there is more going on though.  The experience (and a few others) prompted me to create the `Loader` label for these kinds of issues.

**Comment by rallytime** [comment]
Shoot. It appears that the fix that I backported to the 2015.5 branch didn't make it into the 2015.5.6 release. I thought I had tested it there, but I must have gotten the commit wrong for when the branch was tagged. 

That being said, that comment you linked above @jfindlay seems to have something else going on. I tested the sample state the original user provided. I was able to confirm 2015.5.6 being broken, but the HEAD of 2015.5 is fixed. I responded with my findings.
